src/run/run.c

```diff
@@ -1521,11 +1628,74 @@ static int run_context_update(RunContext *c, const char *path) {
 }
 
 static int on_properties_changed(sd_bus_message *m, void *userdata, sd_bus_error *error) {
-        RunContext *c = ASSERT_PTR(userdata);
+        return run_context_update(ASSERT_PTR(userdata));
+}
 
-        assert(m);
+static int on_disconnected(sd_bus_message *m, void *userdata, sd_bus_error *error) {
+        int r;
+
+        /* If our connection gets terminated, then try to reconnect. This might be because the service
+         * manager is being restarted. Handle this gracefully. */
+
+        log_info("Got disconnected from bus connection. Trying to reconnect...");
+        r = run_context_reconnect(ASSERT_PTR(userdata));
+        if (r < 0)
+                return r;
+
+        return 0;
+}
+
+static int run_context_attach_bus(RunContext *c, sd_bus *bus) {
+        int r;
+
+        assert(c);
+        assert(bus);
+
+        assert(!c->bus);
+        assert(!c->match_properties_changed);
+        assert(!c->match_disconnected);
 
-        return run_context_update(c, sd_bus_message_get_path(m));
+        c->bus = sd_bus_ref(bus);
+
+        r = sd_bus_match_signal_async(
+                        c->bus,
+                        &c->match_properties_changed,
+                        "org.freedesktop.systemd1",
+                        c->bus_path,
+                        "org.freedesktop.DBus.Properties",
+                        "PropertiesChanged",
+                        on_properties_changed, NULL, c);
+        if (r < 0)
+                return log_error_errno(r, "Failed to request properties changed signal match: %m");
```

Comment:
Nit, in other places where we subscribe to this signal we tend to log about the failure and use actual signal name, e.g. "PropertiesChanged". Maybe something to consider...